File: src/main/java/com/dv/telegram/config/BotSettingUtils.java
package com.dv.telegram.config;

import com.dv.telegram.WikiBotConfig;
import org.apache.commons.lang3.StringUtils;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class BotSettingUtils {

    private BotSettingUtils() {
    }

    public static Map<String, BotSetting<?>> fillSettingsMap(WikiBotConfig config) {
        return fillSettingsMap(config.getSettings());
    }

    public static Map<String, BotSetting<?>> fillSettingsMap(Map<String, String> settings) {
        List<BotSetting<?>> allSettings = BotSetting.getAllSettings();

        Map<String, BotSetting<?>> botSettings = new HashMap<>();

        for (Map.Entry<String, String> entry : settings.entrySet()) {
            String name = entry.getKey();
            String value = entry.getValue();

            if (StringUtils.isBlank(name)) {
                throw new SettingValidationException("Setting name cannot be empty.");
            }

            if (StringUtils.isBlank(value)) {
                throw new SettingValidationException("Setting value cannot be empty.");
            }

            BotSetting<?> setting = getBotSetting(allSettings, name);
            setting.setValue(value);

            if (botSettings.containsKey(name)) {
                throw new SettingValidationException(String.format("Duplicate setting: \"%s\".", name));
            }

            botSettings.put(name, setting);
        }

        // todo: validate that all settings are present in the config

        return botSettings;
    }

    private static BotSetting<?> getBotSetting(List<BotSetting<?>> allSettings, String settingName) {
        return allSettings
            .stream()
            .filter(setting -> setting.getName().equals(settingName))
            .findFirst()
            .orElseThrow(() -> new SettingValidationException(String.format("Unknown setting name: \"%s\".", settingName)));
    }
}

File: src/main/java/com/dv/telegram/config/BotSettingUtils.kt
package com.dv.telegram.config

import com.dv.telegram.WikiBotConfig

object BotSettingUtils {
    @JvmStatic
    fun fillSettingsMap(config: WikiBotConfig): Map<String, BotSetting<*>> {
        return fillSettingsMap(config.getSettings())
    }

    @JvmStatic
    fun fillSettingsMap(settings: Map<String?, String?>): Map<String, BotSetting<*>> {
        val botSettings = mutableMapOf<String, BotSetting<*>>()

        for ((name, value) in settings) {
            if (name.isNullOrBlank()) {
                throw SettingValidationException("Setting name cannot be empty.")
            }

            if (value.isNullOrBlank()) {
                throw SettingValidationException("Setting value cannot be empty.")
            }

            val setting = getBotSetting(BotSetting.allSettings, name)
            setting.setValue(value)

            if (botSettings.containsKey(name)) {
                throw SettingValidationException("Duplicate setting: \"${name}\".")
            }

            botSettings[name] = setting
        }

        // todo: validate that all settings are present in the config
        return botSettings
    }

    private fun getBotSetting(allSettings: List<BotSetting<*>>, settingName: String): BotSetting<*> {
        return allSettings
            .firstOrNull { it.name == settingName }
            ?: throw SettingValidationException("Unknown setting name: \"${settingName}\".")
    }
}
